### PR TITLE
Laser Meteor Miner

### PR DESCRIPTION
**Pull request by zTotto**
Adds a proper miner for Blood Magic Meteors since the current solutions are somewhat lacky or annoying to setup. This multi mines all the blocks dropped by the meteor with fortune 3 but it can be easily modified by adding tiers.

https://github.com/user-attachments/assets/ac501ad7-e092-4f88-b9e4-a4d54d944ae5

The multiblock has a max range of 24 blocks but it gets dynamically lowered based on the meteorite size. It has a idle state when there's no meteor above the reference block, which is 32 blocks over the center of the multi's highest block.

https://github.com/user-attachments/assets/abc4172d-6439-4407-8ed3-b04234cc0fe1

The multi also renders a laser from the top block of the coil line up to the upmost layer of the current meteor (the range of the laser is dynamic, the whole rendering can be disabled with a screwdriver).

If needed i could also add some kind of parallels and tiers, i was thinking about mining drones or liquids to cool down the lasers.
If the structure also needs some edits that's a quick edit, for example at the moment all the glass in the multi can be any tier but that's up to balancing, like it is in many multis with glass like the EEC or the MEBF.

(All the code was written after The Great Renaming™)

### CODE REVIEW

**Comment by OneEyeMaker** [comment]
Is the output of this multi comparable to the output of the Ore Drilling Rig on naturally generated ores?

**Comment by EnderProyects** [comment]
Could you make that the bottom blocks of the multi sits on top of the retual or fit around it instead on being floating on top of it ( idk if could be made that the top blocks of the ritual has to be part of the multi)

**Comment by zTotto** [comment]
> Is the output of this multi comparable to the output of the Ore Drilling Rig on naturally generated ores?

I tried on manually placed ores and the output was 1 raw ore per ore, i'll now try with meteors leaving just 1 block but i think it behaves exactly like the drill so it does apply fortune on naturally generated ores, i'm using the getDrops() function with fortune 3

**Comment by zTotto** [comment]
> Could you make that the bottom blocks of the multi sits on top of the ritual or fit around it instead on being floating on top of it ( idk if could be made that the top blocks of the ritual has to be part of the multi)

I can merge the structure with the ritual, i'll check rn if i can directly use the ritual stones as part of the multi

**Comment by zTotto** [comment]
> > Is the output of this multi comparable to the output of the Ore Drilling Rig on naturally generated ores?
> 
> I tried on manually placed ores and the output was 1 raw ore per ore, i'll now try with meteors leaving just 1 block but i think it behaves exactly like the drill so it does apply fortune on naturally generated ores, i'm using the getDrops() function with fortune 3

It does apply fortune 3 (in this case, since i wrote that atm) to mined blocks

**Comment by zTotto** [comment]
Something like this plus the satellite this above?
immagine

**Comment by EnderProyects** [comment]
Yeah, just check that it works and that you can place the automation ( spoiler there is no space for the dinamism tablet in order to work xd)

**Comment by zTotto** [comment]
> Yeah, just check that it works and that you can place the automation ( spoiler there is no space for the dinamism tablet in order to work xd)

yeah i could lower the first layer so that automation can be done oopsie ahah

**Comment by zTotto** [comment]
Could also do something like this?
immagine
and above the usual structure
(tbh i like this better)

**Comment by EnderProyects** [comment]
Is your multi, so what you want xd, I just request the change so it doesnt float in the midle of it

**Comment by zTotto** [comment]
Ok i ended up making this, it looks cleaner, ill check if the structure can support BM blocks
immagine

**Comment by EnderProyects** [comment]
Can we have this merged for testing in nightlys? Also idk if the controller gas a recipe

**Comment by Dream-Master** [comment]
> Can we have this merged for testing in nightlys? Also idk if the controller has a recipe

its already in but i make a new nightly

**Comment by zTotto** [comment]
> Can we have this merged for testing in nightlys? Also idk if the controller has a recipe
Atm the controller and the beacon don't have a recipe

**Comment by Dream-Master** [comment]
@zTotto add some so we can check it

**Comment by zTotto** [comment]
> Also you said the fortune could be tiered? What about having an item in the controller or input bus that gives the level of fortune? Like: Nothing: no fortune Blood pickaxe: fortune 1 Thaumcraft pickaxe: fortune 2 Botania Terra shatterer: fortune 3 ( just an idea that represents the usage of pickaxes in arcane bores)

Yeah i like the idea but at the same time i'd go for something more tech-related since it's a laser, i'll think of something

**Comment by zTotto** [comment]
I added the missing recipes @Dream-Master

File: src/main/java/gregtech/common/tileentities/machines/multi/MTEMeteorMiner.java [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+        for (int z = -currentRadius; z <= (currentRadius - opposite); z++) {
+            int currentZ = this.zStart + z;
+            if (!getBaseMetaTileEntity().getWorld()
+                .isAirBlock(currentX, currentY, currentZ)) {
+                Block target = getBaseMetaTileEntity().getBlock(currentX, currentY, currentZ);
+                if (target.getBlockHardness(getBaseMetaTileEntity().getWorld(), currentX, currentY, currentZ) > 0) {
+                    final int blockMeta = getBaseMetaTileEntity().getMetaID(currentX, currentY, currentZ);
+                    addToOutput(
+                        target
+                            .getDrops(getBaseMetaTileEntity().getWorld(), currentX, currentY, currentZ, blockMeta, 3));
```
Comment (OneEyeMaker) [inline src/main/java/gregtech/common/tileentities/machines/multi/MTEMeteorMiner.java:419]: You shouldn't always mine blocks with fortune. I think that logic of Ore Drilling Rig is preferable: https://github.com/GTNewHorizons/GT5-Unofficial/blob/master/src/main/java/gregtech/common/tileentities/machines/multi/MTEOreDrillingPlantBase.java#L256

Comment (zTotto) [reply src/main/java/gregtech/common/tileentities/machines/multi/MTEMeteorMiner.java]: I added a tier variable, it's temporarily set to 0 but i'll make something up about tiers

**Comment by zTotto** [comment]
Should i remove the addToOutput function now that there's a merge output stack logic?

**Comment by HoleFish** [comment]
if you merge output just for display, yes; for saving memory, no

**Comment by Dream-Master** [comment]
@zTotto please run spotless

**Comment by chochem** [comment]
this should certainly be frozen.

Regarding the contents: They were actually advised against making this before hand but ignored that. It is not clear why this should exist. MBMs work just fine on all ore meteors, arcane bores can be set up for the others. The purpose just seems to be to buff no-rocket further where planets and rockets become less and less relevant in the pack.

**Comment by zTotto** [comment]
As i wrote in the discord, i wanted to do it nonetheless for myself since i don't like arcane bores for this purpose (i used them and i had to use like 50 of them) and eventually it looked good and i thought someone else might like it so i wanted to make a PR, i didn't think of it as a way to skip planets and rockets, i just wanted a good way to mine meteors with a GT multi 😅.
That said i'm totally fine with the PR being rejected, as i said i made it for myself but i thought it might be appreciated by someone else so i wanted to share it

**Comment by Dream-Master** [comment]
@zTotto @chochem instead of reject and not make planet skipable we can think of a mechanism recipe to mine meteors but still need to visit planets fro people not like magic stuff. What about the idea found some schematics for this multi in space? Not sure this multi support different tiers.

**Comment by Dream-Master** [comment]
I mean like the Astro Miner from GC

**Comment by zTotto** [comment]
Atm i only added fortune tiers, what i can do to avoid planet skipping is:
a) Change the recipe to use a certain scheme only found in a certain planet
b) Make the multi need a certain scheme in a selected input bus

**Comment by FourIsTheNumber** [comment]
I see no reason to add tiers to it. The meteors are what allow skipping, nothing to do with this multi. The meteors can already be automatically mined in several different ways, which is what makes this strange.

**Comment by Dream-Master** [comment]
if it has only one tier a schematic would be nice. Sure it not need multiple tier. @zTotto In which Tier this multi is atm ?

**Comment by zTotto** [comment]
It's tiered at LuV, it also needs superconductor coils so it requires Trinium

**Comment by zTotto** [comment]
I could directly change the recipe for the Meteor Miner by adding the Astro Miner schematic in the recipe, how does it work for external items? for example the scanner uses
GTModHandler.getModItem(GalacticraftMars.ID, "item.schematic", 1L, 2))
can i use it too or is it better if the recipe is in another mod?

**Comment by Dream-Master** [comment]
> I could directly change the recipe for the Meteor Miner by adding the Astro Miner schematic in the recipe, how does it work for external items? for example the scanner uses GTModHandler.getModItem(GalacticraftMars.ID, "item.schematic", 1L, 2)) can i use it too or is it better if the recipe is in another mod?

The astro miner tier is below Luv so this change only required Tier 2 Rocket.

**Comment by zTotto** [comment]
A new schematic is doable for me, i can also make the art for it, i just don't know how to add it to planet chests atm but i can have a look ahah
I could make more tiers which add more fortune or more "parallels" (as in more rows per tier) but imo 2 tiers would be enough.
Which one would be better in your opinion?

**Comment by zTotto** [comment]
This would be the Tier 2 (it requires fusion coils, neutronium frames + mining casings and black plutonium frames + mining casings)
immagine
tomorrow i'll make it into the code and i can make a second schematic to upgrade it to tier 2 (like the multiblock macerator)

**Comment by zTotto** [comment]
I made all the code for everything, what's left to make is the items textures and i'm on it rn

**Comment by S4mpsa** [comment]
Why does this need a schematic? It doesn't gate anything unlike the Rockets, so having a schematic for it is only an annoyance. Unless the reason for having a schematic is so people doing skyblock cannot have access to this, which would be quite targeted considering that's where meteors are the most useful.

**Comment by boubou19** [comment]
We want people doing this not able to skip rockets. That's all.

**Comment by S4mpsa** [comment]
How does adding schematics to this accomplish that, when the player can just use any of the already existing ways of mining meteors?

**Comment by boubou19** [comment]
by doing this i mean using the multi. People can use arcane bores if they want to skip the rockets.

**Comment by Miklebe** [comment]
I can understand the intent behind locking it so that a lot of people wouldn't just start using meteors instead of rockets, but considering they are not consumable schematics, unlike Galacticraft rockets, hunting for another dungeon and praying that it's not bugged and has boss/chest is miserable. Unless the conversion recipe is added like with Astro Miner, so you can first scan the rocket schematic and then make however much you want of them and convert to mining ones.
But also like:
- Arcane bore is an HV recipe (locked behind fusion), they don't drop more raw ores with fortune (or i had no 0 luck).
- Ore Drilling 4 with the lowest processing time is still measly 20 blocks per second (which is around UV+ with double hatches). It took 30+ minutes to mine single meteor, which takes 30 seconds using tier 2 of Meteor Miner. You can just spam more miners, but isn't it more about using less multis at higher tiers to reduce lag?
It's already LuV/ZPM locked, Blood Magic itself for the most people is not the nicest mod (or so i heard when suggesting meteors for resource gen), and it still doesn't even have all ores (yet?), like mytryl, which is also crucial for magic... (or at least makes ichorium less miserable and efficient)

I have not played skyblock yet, but i plan on doing so and would immensely love this being an option for ore mining, even if schematic is still kept, but at least made some considerable gate behind it (it being LuV is already pretty high, unless schematic will use some annoying materials from different mods, or X amount of Bound Diamonds/Blood Infused Diamond blocks)


Anyway, on bug report. Seems like the multi doesn't work with otherworld blocks (netherack, endstone), and still only mines them like regular blocks.
image
javaw_vMPejtKDDs
Meteor Miner only mined 184 raw ores, while MB miner got 876 (3rd from the last slot in chest)
We can get around 360 crushed ores from raw mined ores, when accounting for the nature of fortune, almost fits twice as less calculation of not using double drops from otherwordly ore blocks.

**Comment by zTotto** [comment]
First of all thanks for the feedback

Atm im on vacation and I'll be back on Tuesday, that said when I'm back I'll check what you said about otherworld blocks, tho I tested it with some endstone meteors (with fortune 0) and i was getting the ores listed in the meteor recipe so i have no idea, I'll still have a look.

About the schematic: I was planning on making players able to duplicate it just like the other schematics so no random running for more than 1 miner

**Comment by Miklebe** [comment]
Hmmm, on the second notice, yes, it does seem to double.
With no fortune on tier 1 miner it gave me this much
image
Then using Terra Shatterer i get this result
javaw_MMyiPQrH6e
I will probably run a few tests, but seeing how it's not higher than the tier 2 i posted begore, i have a feeling maybe fortune is borked?
Does it matter if Terra Shatterer has mana or not?
A few minute later... Tested with Pickaxe of the core, already seems higher yield. I will try to run a few tests using AE for accurate numbers, but seems like either Terra Shatterer has NBT problems which only accepts fresh non mana version, or tier 2 has no fortune bonus.
javaw_MXks11jgaT

**Comment by zTotto** [comment]
Mind that ATM I'm only checking the name of the item since I didn't know if I could check for extra mods items lmao, ofc that was temporary but I'll add a fortune indicator on the scanned info or waila

**Comment by zTotto** [comment]
I think i found the problem about the "Fortune Dilemma", i'll see what i can do since apparently the getDrops method i'm using (the same as the GT miner) doesn't apply fortune to GT ores

**Comment by chochem** [comment]
that is intentional though.

**Comment by zTotto** [comment]
Yup, i was thinking about adding another bonus to T2/adding magic pickaxes to the T1

**Comment by Caedis** [comment]
Raw ore fortuning is only with manual mining. Full stop.

**Review by chochem — CHANGES_REQUESTED** [review]
the dev team seems to be united against this:
- its a gt multi replacing a magic task
- it takes away the thought/logistics involved with arcane bore
- raw ores are only for manual mining
- meteors are quite strong and shouldnt be buffed
- the plan/design wasnt approved
- we are in feature freeze

dream seems a bit more open then the rest but also said he doesnt approve "until it fit[s] into gtnh progression"

**Comment by boubou19** [comment]
a small note: we have GT multis replacing magic tasks, like the essentia smelter or researcher. It's not necessarily a bad thing, especially when magic mods are oftenly a big pain to deal with. If base arcane bores were less of a pain to use, the gap between the bore and this multi wouldn't be perceived as high

**Comment by zTotto** [comment]
Raw ores can be easily changed, i could make it output directly the ore block (like the void miner does) or the crushed ore, although thaumcraft bores mining meteors still drop raw ores

Imo this is not a meteor buff but more of a QoL since arcane bores are painful to setup, there's no "thought" or anything like that (from my experience i just spammed them in a grid to cover all the meteor possible space, and that was just annoying, not a logistic challenge or anything like that)

Im totally fine with the feature freeze, im not in a hurry or anything

**Comment by boubou19** [comment]
> Imo this is not a meteor buff but more of a QoL since arcane bores are painful to setup, there's no "thought" or anything like that (from my experience i just spammed them in a grid to cover all the meteor possible space, and that was just annoying, not a logistic challenge or anything like that)

If we strictly compare arcane bores and your multi, one is imperfect solution, the other is a perfect solution. So because you remove all any possible tradeoffs you had to do in order to exploit the arcane bores (how should you control it, do you want to spam this setup, etc). The issue is that while we agreed among the team to yeet massive tedium tasks, we aren't going to remove **all** the tedium from the game, especially if it induces thinking processes. QoL is a slippery slope, as what is acceptable for someone is annoying for the others, so with this logic of QoL, we can end up removing any challenge by turning anything into black boxes.

Also from what you said you only considered arcane bores, but there are other setups to explore to understand the gist of every automation, check their pros and cons. Only after that you can propose a solution.

 Right know, this multi has potential to become THE way of progressing, ignoring miners in general. That would be annoying as ore meteors are supposed to be a side way of gathering ores.

**Comment by Dream-Master** [comment]
Why not using Esentia as drill fluid and change the amimation to match more like a Arcane Bore 2.0 ?

**Comment by zTotto** [comment]
I was thinking i could change some things, please tell me if you like them:

- Make the multi consume centivis and some essentia (we can decide which one later, maybe perfodio or something like that), centivis or essentia scale with EU tier

- Remove fortune and replace fortune tiers with efficiency tiers which lower the centivis or essentia required for each cycle, meaning that better pickaxe are more "magic attuned" and make the multi more "magic energy" efficient

**Comment by zTotto** [comment]
> > I was thinking i could change some things, please tell me if you like them:
> > 
> > * Make the multi consume centivis and some essentia (we can decide which one later, maybe perfodio or something like that), centivis or essentia scale with EU tier
> > * Remove fortune and replace fortune tiers with efficiency tiers which lower the centivis or essentia required for each cycle, meaning that better pickaxe are more "magic attuned" and make the multi more "magic energy" efficient
> 
> 1: in my opinion centivis is too easy to get a stable supply of for this to be realistic. Maybe the Astral Essentia or Space Essentia from GT for Thaumcraft is a better idea. That way that Essentia gets a nice use, from being useless.
> 
> 2: why even have a fortune for the ores when that explicitly is set to be manual mining only. My opinion is to not give fortune for this, it’s powerful enough already without it.

I didn't know about the fortune stuff, i did it because i was looking at other miners and they had fortune (ore drills), now as i said i can remove it no problem, what do you think about the essentia efficiency bonus?

**Comment by Ethryan** [comment]
In my opinion it should need 1 of the essentia types to even run, maybe supply a different harder to get on for a small 15 - 20 %, efficiency bonus.

But honestly, I’m crap at balancing stuff, so amount eaten and bonus given probably needs to be tweaked.

**Comment by NotAPenguin0** [comment]
Feeding essentia doesn't really solve the issue at hand. Essentia is completely free to supply at this point in the pack, and the meteor miner will still easily be the best option for mining meteors without any major tradeoffs or downsides. Especially adding *another* 15-20% bonus on top would make it even more broken. The meteor miner should in my opinion be balanced against the power level of MBMs but specialized for mining meteors. This means
- Similar speed to MBM
- Similar power draw to MBM
- Same yield as MBM
- Slightly higher material cost for the convenience of it being specialized for meteors
- Specialized for meteors, meaning it won't leave random blocks unmined like MBMs do.

This way it's a clear alternative to all other options, instead of simply being superior in every way.

**Comment by zTotto** [comment]
At the moment the last 2 points are already true since it costs quite a lot more than an MBM (i could switch the T3 miner needed to craft it with a T4 miner)
Removing fortune would make the yield less than an MBM since that applies fortune to crashed ores and this wouldn't, fine by me
Power draw is not a problem, just a quick edit and its good
The speed could be balanced so that it mines 1 block at a time instead of a row, i could make the T2 mine a row at a time instead (since that's more advanced progression wise)

**Comment by NotAPenguin0** [comment]
Just use the same drop code as MBMs, no need to complicate it further than that. I think it was already said that fortuned raw ores are not supposed to be used in automatic miners.

**Comment by zTotto** [comment]
I'll update everything, im just busy these days (got a new dog hehe), probably will do everything before next week

File: src/main/java/gregtech/common/tileentities/machines/multi/MTEMeteorMiner.java [hunk 2]
```diff
@@ -0,0 +1,764 @@
...
+            if (!getBaseMetaTileEntity().getWorld()
+                .isAirBlock(xStart, yStart, zCoord)) {
+                break;
+            }
+            delta++;
+        }
+        currentRadius -= delta;
+    }
+
+    protected void setElectricityStats() {
```
Comment (NotAPenguin0) [inline src/main/java/gregtech/common/tileentities/machines/multi/MTEMeteorMiner.java:703]: This should really use the overclock calculator utility. In general, you should try to swap a lot of these things to use `ProcessingLogic` instead if possible.

**Review by NotAPenguin0 — CHANGES_REQUESTED** [review]

File: src/main/java/gregtech/common/tileentities/machines/multi/MTEMeteorMiner.java (hunk 2 above)
Comment (zTotto) [reply src/main/java/gregtech/common/tileentities/machines/multi/MTEMeteorMiner.java:704]: Ok i added a calculator, it uses the same values as the T1 MBM

**Comment by serenibyss** [comment]
At the moment we don't want to take this feature. Feel free to bring it up for further discussion in discord @zTotto

**Pull request closed without merging**